lib/flowViewport/sceneIndex/fvpPruningSceneIndex.cpp

```diff
@@ -0,0 +1,344 @@
+{
+    return prim.primType == HdPrimTypeTokens->capsule;
+}
+
+bool _ConesFilterHandler(const HdSceneIndexBaseRefPtr& sceneIndex, const SdfPath& primPath, const HdSceneIndexPrim& prim)
+{
+    return prim.primType == HdPrimTypeTokens->cone;
+}
+
+bool _CubesFilterHandler(const HdSceneIndexBaseRefPtr& sceneIndex, const SdfPath& primPath, const HdSceneIndexPrim& prim)
+{
+    return prim.primType == HdPrimTypeTokens->cube;
+}
+
+bool _CylindersFilterHandler(const HdSceneIndexBaseRefPtr& sceneIndex, const SdfPath& primPath, const HdSceneIndexPrim& prim)
+{
+    return prim.primType == HdPrimTypeTokens->cylinder;
+}
+
+bool _SpheresFilterHandler(const HdSceneIndexBaseRefPtr& sceneIndex, const SdfPath& primPath, const HdSceneIndexPrim& prim)
+{
+    return prim.primType == HdPrimTypeTokens->sphere;
+}
+
+bool _NurbsCurvesFilterHandler(const HdSceneIndexBaseRefPtr& sceneIndex, const SdfPath& primPath, const HdSceneIndexPrim& prim)
+{
+    return prim.primType == HdPrimTypeTokens->nurbsCurves;
+}
+
+bool _NurbsPatchesFilterHandler(const HdSceneIndexBaseRefPtr& sceneIndex, const SdfPath& primPath, const HdSceneIndexPrim& prim)
+{
+    return prim.primType == HdPrimTypeTokens->nurbsPatch;
+}
+
+} // namespace
+
+namespace FVP_NS_DEF {
+
+PruningSceneIndexRefPtr PruningSceneIndex::New(const HdSceneIndexBaseRefPtr &inputSceneIndex)
+{
+    return TfCreateRefPtr(new PruningSceneIndex(inputSceneIndex));
+}
+
+PruningSceneIndex::PruningSceneIndex(HdSceneIndexBaseRefPtr const &inputSceneIndex) : 
+    HdSingleInputFilteringSceneIndexBase(inputSceneIndex), 
+    InputSceneIndexUtils(inputSceneIndex)
+{
+}
+
+void PruningSceneIndex::AddExcludedSceneRoot(const PXR_NS::SdfPath& sceneRoot)
+{
+    _excludedSceneRoots.emplace(sceneRoot);
+}
+
+bool PruningSceneIndex::_IsExcluded(const PXR_NS::SdfPath& primPath) const
+{
+    return _HasAncestorInclusiveInContainer(primPath, _excludedSceneRoots);
+}
+
+bool PruningSceneIndex::_PrunePrim(const SdfPath& primPath, const HdSceneIndexPrim& prim, const TfToken& pruningToken) const
+{
+    if (_IsExcluded(primPath)) {
+        return false;
+    }
+    using FilterHandler = std::function<bool(const HdSceneIndexBaseRefPtr&, const SdfPath&, const HdSceneIndexPrim&)>;
+    static std::map<TfToken, FilterHandler> filterHandlers = {
+        { FvpPruningTokens->meshes, _MeshesFilterHandler },
+        { FvpPruningTokens->capsules, _CapsulesFilterHandler },
+        { FvpPruningTokens->cones, _ConesFilterHandler },
+        { FvpPruningTokens->cubes, _CubesFilterHandler },
+        { FvpPruningTokens->cylinders, _CylindersFilterHandler },
+        { FvpPruningTokens->spheres, _SpheresFilterHandler },
+        { FvpPruningTokens->nurbsCurves, _NurbsCurvesFilterHandler },
+        { FvpPruningTokens->nurbsPatches, _NurbsPatchesFilterHandler }
+    };
+    return filterHandlers[pruningToken](GetInputSceneIndex(), primPath, prim);
```

Comment:
The basic idea of this scene index is to associate TfTokens with predicates that return whether or not to filter out a given prim, so we can separate out the filtering logic per-filter. Currently we don't really need special checks, so the handlers just use the prim type. 

This mechanism could potentially be exposed externally if there is a need (e.g. to allow plugin developers to add their own filters), though for now there is not so I've kept this internal to the class. 

The rest of the implementation of this class mostly consists of tracking which prims are pruned out or not, and sending/blocking corresponding PrimsAdded/PrimsRemoved notifications.